src/input.c

```diff
@@ -333,122 +361,168 @@ void _glfwInputMouseClick(_GLFWwindow* window, int button, int action, int mods)
+    assert(entered == GLFW_TRUE || entered == GLFW_FALSE);
+
     if (window->callbacks.cursorEnter)
-        window->callbacks.cursorEnter((GLFWwindow*) window, entered);
+        window->callbacks.cursorEnter((GLFWwindow *)window, entered);
 }
 
 // Notifies shared code of files or directories dropped on a window
 //
-void _glfwInputDrop(_GLFWwindow* window, int count, const char** paths)
+void _glfwInputDrop(_GLFWwindow *window, int count, const char **paths)
 {
+    assert(window != NULL);
+    assert(count > 0);
+    assert(paths != NULL);
+
     if (window->callbacks.drop)
-        window->callbacks.drop((GLFWwindow*) window, count, paths);
+        window->callbacks.drop((GLFWwindow *)window, count, paths);
 }
 
 // Notifies shared code of a change in the gamepad state.
 // Automatically recalculates the state if the gamepad callback is installed.
-void _glfwInputGamepad(_GLFWjoystick* js)
+void _glfwInputGamepad(_GLFWjoystick *js)
 {
-    const int jid = (int) (js - _glfw.joysticks);
+    const int jid = (int)(js - _glfw.joysticks);
 
-    if (!_glfw.initialized) {
+    if (!_glfw.initialized)
+    {
         return;
     }
 
-    if ((js->mapping != NULL)  && (_glfw.callbacks.gamepad_state)) {
+    if ((js->mapping != NULL) && (_glfw.callbacks.gamepad_state))
+    {
         GLFWgamepadstate state;
-        if (glfwGetGamepadState(jid, &state)) {
-          _glfw.callbacks.gamepad_state(jid, state.buttons, state.axes);
+        if (glfwGetGamepadState(jid, &state))
+        {
+            _glfw.callbacks.gamepad_state(jid, state);
         }
     }
 }
 
 // Notifies shared code of a joystick connection or disconnection
 //
-void _glfwInputJoystick(_GLFWjoystick* js, int event)
+void _glfwInputJoystick(_GLFWjoystick *js, int event)
 {
-    if (_glfw.callbacks.joystick) {
-        const int jid = (int) (js - _glfw.joysticks);
-        _glfw.callbacks.joystick(jid, event);
-    }
-}
+    assert(js != NULL);
+    assert(event == GLFW_CONNECTED || event == GLFW_DISCONNECTED);
+
+    if (event == GLFW_CONNECTED)
+        js->connected = GLFW_TRUE;
+    else if (event == GLFW_DISCONNECTED)
+        js->connected = GLFW_FALSE;
 
+    if (_glfw.callbacks.joystick)
+        _glfw.callbacks.joystick((int)(js - _glfw.joysticks), event);
+}
 
 // Notifies shared code of the new value of a joystick axis
 //
-void _glfwInputJoystickAxis(_GLFWjoystick* js, int axis, float value)
+void _glfwInputJoystickAxis(_GLFWjoystick *js, int axis, float value)
 {
-    if (js->axes[axis] != value) {
-        const int jid = (int) (js - _glfw.joysticks);
+    assert(js != NULL);
+    assert(axis >= 0);
+    assert(axis < js->axisCount);
+
+    if (js->axes[axis] != value)
+    {
+        const int jid = (int)(js - _glfw.joysticks);
 
         js->axes[axis] = value;
         if (_glfw.callbacks.joystick_axis)
             _glfw.callbacks.joystick_axis(jid, axis, value);
         _glfwInputGamepad(js);
+    } else {
+        js->axes[axis] = value;
```

Comment:
merged with upstream changes and meshed the logic together, is mostly the same structure for all input methods on the joystick